pkg/provisioners/gcppubsub/channel/reconcile.go

```diff
@@ -0,0 +1,325 @@
+	if reconcileErr != nil {
+		logging.FromContext(ctx).Info("Error reconciling Channel", zap.Error(reconcileErr))
+		// Note that we do not return the error here, because we want to update the Status
+		// regardless of the error.
+	}
+
+	if err = util.UpdateChannel(ctx, r.client, c); err != nil {
+		logging.FromContext(ctx).Info("Error updating Channel Status", zap.Error(err))
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, reconcileErr
+}
+
+// shouldReconcile determines if this Controller should control (and therefore reconcile) a given
+// Channel. This Controller only handles gcp-pubsub channels.
+func (r *reconciler) shouldReconcile(c *eventingv1alpha1.Channel) bool {
+	if c.Spec.Provisioner != nil {
+		return ccpcontroller.IsControlled(c.Spec.Provisioner)
+	}
+	return false
+}
+
+func (r *reconciler) reconcile(ctx context.Context, c *eventingv1alpha1.Channel) error {
+	c.Status.InitializeConditions()
+
+	// We are syncing four things:
+	// 1. The K8s Service to talk to this Channel.
+	// 2. The Istio VirtualService to talk to this Channel.
+	// 3. The GCP PubSub Topic (one for the Channel).
+	// 4. The GCP PubSub Subscriptions (one for each Subscriber of the Channel).
+
+	// Regardless of what we are going to do, we need GCP credentials to do it.
+	gcpCreds, err := pubsubutil.GetCredentials(ctx, r.client, r.defaultSecret, r.defaultSecretKey)
+	if err != nil {
+		logging.FromContext(ctx).Info("Unable to generate GCP creds", zap.Error(err))
+		return err
+	}
+
+	if c.DeletionTimestamp != nil {
+		// K8s garbage collection will delete the K8s service and VirtualService for this channel.
+		err = r.deleteSubscriptions(ctx, c, gcpCreds, r.defaultGcpProject)
+		if err != nil {
+			return err
+		}
+		err = r.deleteTopic(ctx, c, gcpCreds, r.defaultGcpProject)
+		if err != nil {
+			return err
+		}
+		util.RemoveFinalizer(c, finalizerName)
+		return nil
+	}
+
+	// We use a finalizer to ensure the GCP PubSub Topic and Subscriptions are deleted.
+	util.AddFinalizer(c, finalizerName)
+
+	err = r.createK8sService(ctx, c)
+	if err != nil {
+		return err
+	}
+
+	err = r.createVirtualService(ctx, c)
+	if err != nil {
+		return err
+	}
+
+	topic, err := r.createTopic(ctx, c, gcpCreds, r.defaultGcpProject)
+	if err != nil {
+		return err
+	}
+
+	err = r.createSubscriptions(ctx, c, gcpCreds, r.defaultGcpProject, topic)
+	if err != nil {
+		return err
+	}
+
+	c.Status.MarkProvisioned()
+	return nil
+}
+
+func (r *reconciler) createK8sService(ctx context.Context, c *eventingv1alpha1.Channel) error {
+	svc, err := util.CreateK8sService(ctx, r.client, c)
+	if err != nil {
+		logging.FromContext(ctx).Info("Error creating the Channel's K8s Service", zap.Error(err))
+		return err
+	}
+
+	// Check if this Channel is the owner of the K8s service.
+	if !metav1.IsControlledBy(svc, c) {
+		logging.FromContext(ctx).Warn("Channel's K8s Service is not owned by the Channel", zap.Any("channel", c), zap.Any("service", svc))
```

Comment:
I created https://github.com/knative/eventing/issues/648 to track this.

Reply:
My understanding is that this is what happens.

1. In the user code, an HTTP request is made to the `Channel`'s DNS name. 
    - This is injected via the `Channel`'s `status.address`.
    - Currently this is based on the `Channel`'s K8s `Service`. e.g. 'qux-1-channel.default.svc.cluster.local'
    - Where this request is routed isn't important, as long as the user code is happy to send the request.
1. The request is intercepted by the Istio sidecar. Which notices the host header in the request and sees that it matches a `VirtualService`'s `spec.hosts` entry. So it ignores where the request was originally sent and instead routes it to the destination K8s `Service`. It also rewrites the host header to the 'Channel DNS' (e.g. 'qux-1.default.channels.cluster.local').
1. The request is sent to a Pod on the destination K8s `Service`. Which uses the host header to determine which `Channel` this request is being made on.

So, we could easily change `VirtualService` to a generated name. 

We could also change the K8s `Service` to a generated name, as long as we update the `Channel`'s `status.address` to the correct generated DNS. Note that making this change would cause a blip in `Channel`'s when switching from the 'normal' name to the generated one while all `Subscriptions` get the new updated value and the Istio mesh updates. We might be able to do it without any outage by using label selectors to find the `Service`, which we can make match the existing `Service`s (so that existing `Channel` continue to use the nicely named `Service` while all news ones use generated names).